src/EventStore.Core/Services/ClusterStorageWriterService.cs

```diff
@@ -333,23 +343,11 @@ public class ClusterStorageWriterService<TStreamId> : StorageWriterService<TStre
 			}
 		}
 
-		private void OnLogRecordUnframed(BinaryReader reader) {
-			var rawLength = reader.BaseStream.Length;
-
-			if (rawLength >= int.MaxValue)
-				throw new ArgumentOutOfRangeException(
-					nameof(reader),
-					$"Length of stream was {rawLength}");
-
-			var length = (int)rawLength;
-
-			var record = LogRecord.ReadFrom(reader, length: length);
-			long newPos;
-			if (!Writer.Write(record, out newPos))
-				ReplicationFail(
-					"First write failed when writing replicated record: {0}.",
-					"First write failed when writing replicated record: {record}.",
-					record);
+		private void OnTransactionUnframed(ILogRecord[] records) {
+			Writer.OpenTransaction();
+			foreach (var record in records)
+				Writer.WriteToTransaction(record, out _);
```

Comment:
should we call the TryWriteToTransaction instead and call ReplicationFail if it returns false?

Reply:
fixed :+1: